src/java.net.http/share/classes/jdk/internal/net/http/Http1Response.java

```diff
@@ -331,7 +331,7 @@ public <U> CompletableFuture<U> readBody(HttpResponse.BodySubscriber<U> p,
                 );
                 if (cf.isCompletedExceptionally()) {
                     // if an error occurs during subscription
-                    connection.close();
+                    connection.close(cf.exceptionNow());
```

Comment:
`Future.exceptionNow()` (among other things) states that it will throw a `IllegalStateException` if the Future was cancelled. We do have a check for `cf.isCompletedExceptionally()` before invoking this method. However, the `CompletableFuture.isCompletedExceptionally()` states that it returns `true` even in the case of cancellation:

> Returns {@code true} if this CompletableFuture completed exceptionally, in any way. Possible causes include cancellation ...

So I suspect this call to `exceptionNow()` might need a rethink or some additional checks?

Reply:
Good catch. I suspect the case where the CF is cancelled should not occur, but I have added a Utils.exceptionNow method to take care of this case just in case.